### PR TITLE
Output C code from harness generation tool.

### PR DESCRIPTION
**Pull request by NlightNFotis**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/goto-harness/havoc-global-int-02/main.c [hunk 1]
```diff
@@ -3,7 +3,7 @@
 unsigned int x;
 unsigned int y;
 
-unsigned int nondet_int()
+unsigned int some_random_unsigned_integer()
```
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-harness/havoc-global-int-02/main.c]: This was changed because of a `dump_c()` bug; It assumes that `nondet_*` names are available with specific meanings.

File: regression/goto-harness/havoc-global-int-02/test.desc [hunk 2]
```diff
@@ -1,4 +1,4 @@
-CORE
+KNOWNBUG
```
Comment (hannes-steffenhagen-diffblue) [inline regression/goto-harness/havoc-global-int-02/test.desc:1]: We can work around this for now by re-enabling the old behaviour for these, but we've marked them KNOWNBUG for now because we're first trying to get the call-function harness to behave like we want it to.

File: src/goto-harness/common_harness_generator_options.h [hunk 3]
```diff
@@ -18,6 +18,7 @@ Author: Diffblue Ltd.
   "function-pointer-can-be-null"
 #define COMMON_HARNESS_GENERATOR_HAVOC_MEMBER_OPT "havoc-member"
 
+
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/common_harness_generator_options.h]: 🚫 Whitespace

File: src/goto-harness/function_call_harness_generator.cpp [hunk 4]
```diff
@@ -215,6 +215,8 @@ void function_call_harness_generatort::handle_option(
   }
 }
 
+#include <iostream>
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/function_call_harness_generator.cpp]: 🚫 debug code changes

File: src/goto-harness/function_call_harness_generator.cpp [hunk 5]
```diff
@@ -478,7 +480,7 @@ function_call_harness_generatort::implt::declare_arguments(
   for(const auto &parameter : parameters)
   {
     auto argument = allocate_objects.allocate_automatic_local_object(
-      parameter.type(), parameter.get_base_name());
+      remove_const(parameter.type()), parameter.get_base_name());
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/function_call_harness_generator.cpp:481]: The `remove_const` changes should probably all be in one separate commit.

File: src/goto-harness/recursive_initialization.cpp [hunk 6]
```diff
@@ -1002,7 +1021,7 @@ code_blockt recursive_initializationt::build_function_pointer_constructor(
     }
     ++function_pointer_index;
   }
-
+
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/recursive_initialization.cpp]: stray whitespace change

File: src/util/type.h [hunk 7]
```diff
@@ -224,4 +224,8 @@ inline type_with_subtypest &to_type_with_subtypes(typet &type)
     for(type_with_subtypest::subtypest::iterator it=to_type_with_subtypes(type).subtypes().begin(); \
         it!=to_type_with_subtypes(type).subtypes().end(); ++it)
 
+/// Remove const qualifier from type (if any).
+/// Returns type as is if there is no const qualifier.
+typet remove_const(typet type);
```
Comment (thk123) [inline src/util/type.h:229]: :bulb: I guess I'd probably expect this function to recurse... perhaps it should just be a static util function in this place where it is needed?

File: src/goto-harness/recursive_initialization.cpp [hunk 8]
```diff
@@ -690,12 +690,17 @@ code_blockt recursive_initializationt::build_pointer_constructor(
     seen_assign_prev = code_assignt{*has_seen, has_seen_prev};
   }
 
-  const symbol_exprt &local_result =
-    get_fresh_local_typed_symexpr("local_result", type, nullptr_expr);
+  // we want to initialize the pointee as non-const even for pointer to const
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.cpp:712]: :pick: Add why to the comment

File: src/goto-harness/recursive_initialization.cpp [hunk 9]
```diff
@@ -797,9 +823,9 @@ code_blockt recursive_initializationt::build_dynamic_array_constructor(
   const optionalt<irep_idt> &lhs_name)
 {
   PRECONDITION(result.type().id() == ID_pointer);
-  const typet &pointer_type = result.type().subtype();
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.cpp:800]: :pick: This renaming should have gone in a separate commt

File: src/goto-harness/recursive_initialization.cpp [hunk 10]
```diff
@@ -1008,6 +1008,7 @@ code_blockt recursive_initializationt::build_function_pointer_constructor(
 
   const auto function_pointer_selector =
     get_fresh_local_symexpr("function_pointer_selector");
+  body.add(code_declt{function_pointer_selector});
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.cpp:1031]: :question: Is there a test for this bug? That is, a test that passes if and only if this line is present

File: regression/goto-harness/havoc-global-int-02/test.desc (hunk 2 above)
Comment (thk123) [inline regression/goto-harness/havoc-global-int-02/test.desc:1]: :no_entry_sign: As a rule - tests marked as knownbug should reference either a jira ticket or github issue that explains how they should be turned back on - otherwise we might as well delete them. Could you raise a Jira ticket to this effect and put it in the comment section of each desc file

File: src/goto-harness/goto_harness_parse_options.cpp [hunk 11]
```diff
@@ -9,24 +9,58 @@ Author: Diffblue Ltd.
...
+    goto_model.get_symbol_table().begin(),
+    goto_model.get_symbol_table().end(),
+    std::inserter(symbols, symbols.end()),
+    [](const std::pair<const irep_idt, symbolt> &key_value_pair) {
+      return key_value_pair.first;
+    });
+  return symbols;
+}
+
+static void filter_goto_model(
```
Comment (thk123) [inline src/goto-harness/goto_harness_parse_options.cpp:50]: An ideal candidate for a simple unit test

File: src/goto-harness/goto_harness_parse_options.cpp [hunk 12]
```diff
@@ -50,6 +84,7 @@ int goto_harness_parse_optionst::doit()
 
   auto factory_options = collect_generate_factory_options();
 
+  register_language(new_ansi_c_language);
```
Comment (thk123) [inline src/goto-harness/goto_harness_parse_options.cpp]: :no_entry_sign: Could a why comment explaining why this is needed be added here

File: src/goto-harness/goto_harness_parse_options.cpp [hunk 13]
```diff
@@ -87,13 +124,15 @@ int goto_harness_parse_optionst::doit()
...
-    throw system_exceptiont{"failed to write goto program from file '" +
-                            got_harness_config.out_file + "'"};
-  }
+  filter_goto_model(goto_model, goto_model_without_harness_symbols);
+  auto harness_out = std::ofstream{got_harness_config.out_file};
+  dump_c(
+    goto_model.goto_functions,
+    true,
+    true,
+    false,
```
Comment (thk123) [inline src/goto-harness/goto_harness_parse_options.cpp:131]: :pick: Pull these out into named local variables so there is some clue as to what they match up to. E.g. 

```C++
const bool use_system_headers = true;
...
dump_c(
  goto_model.goto_functions, 
  use_system_headers, 
...
```

File: src/goto-harness/recursive_initialization.cpp [hunk 14]
```diff
@@ -472,35 +472,54 @@ std::string recursive_initialization_configt::to_string() const
...
+}
+
+irep_idt recursive_initializationt::get_fresh_global_name(
+  const std::string &symbol_name,
+  const exprt &initial_value) const
+{
+  auto &fresh_symbol = get_fresh_global_symbol(
+    goto_model.symbol_table,
+    symbol_name,
+    signed_int_type(), // FIXME why always signed_int_type???
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.cpp:509]: :question: Noting this FIXME - it does seem odd?

File: src/goto-harness/recursive_initialization.cpp [hunk 15]
```diff
@@ -472,35 +472,54 @@ std::string recursive_initialization_configt::to_string() const
...
-  const std::string &symbol_name,
-  const exprt &initial_value) const
+static symbolt &get_fresh_global_symbol(
+  symbol_tablet &symbol_table,
+  const std::string &symbol_base_name,
+  typet symbol_type,
+  irep_idt mode)
 {
+  source_locationt source_location{};
+  source_location.set_file(CPROVER_PREFIX "harness.c");
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.cpp]: :pick: magic string "harness" should probably be turned into a macro or constexpr

File: src/goto-harness/recursive_initialization.h [hunk 16]
```diff
@@ -22,6 +22,7 @@ Author: Diffblue Ltd.
 #include "function_harness_generator_options.h"
 #include "goto_harness_generator.h"
 
+#define GOTO_HARNESS_PREFIX "GOTO_HARNESS"
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.h]: :grey_question: Why not 
```suggestion
#define GOTO_HARNESS_PREFIX CPROVER_PREFIX "_GOTO_HARNESS"
```

File: src/goto-harness/recursive_initialization.h [hunk 17]
```diff
@@ -103,12 +103,14 @@ class recursive_initializationt
   /// \return the symbol expression for the `free` function
   symbol_exprt get_free_function();
 
-  bool is_initialization_allowed(const symbolt &symbol)
+  static bool is_initialization_allowed(const symbolt &symbol)
```
Comment (thk123) [inline src/goto-harness/recursive_initialization.h:106]: :pick: Great candidate for a unit test

**Review by thk123 — COMMENTED** [review]
Yeah this looks mostly good to go to me. Just a few minor :no_entry_sign: comments, and a couple of :question: I'd like answered, but otherwise :+1: and of course figuring out the CI issue!

File: src/goto-harness/recursive_initialization.h (hunk 16 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.h]: @thk123 anything starting with `CPROVER_PREFIX` is filtered out and ignored by `dump_c`, that's the only reason we added a goto harness specific prefix.

File: src/util/type.h (hunk 7 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/type.h:229]: I was modelling this after [`std::remove_const`](https://en.cppreference.com/w/cpp/types/remove_cv) which doesn't recurse

File: src/goto-harness/recursive_initialization.cpp (hunk 10 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.cpp:1031]: The existing function pointer tests in goto harness wouldn't really work (anymore) without this.

File: src/goto-harness/recursive_initialization.cpp (hunk 14 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.cpp:509]: @thk123 basically this is because of a somewhat odd design decision that was made at some point because someone was adding a bunch of global variables that were all integers apparently... I don't like it, but I considered it out of scope to actually fix this in this PR.

File: src/goto-harness/Makefile [hunk 18]
```diff
@@ -16,6 +16,10 @@ OBJ += \
   ../linking/linking$(LIBEXT) \
   ../json/json$(LIBEXT) \
   ../json-symtab-language/json-symtab-language$(LIBEXT) \
+  ../ansi-c/ansi-c$(LIBEXT) \
+  ../cpp/cpp$(LIBEXT) \
+  ../goto-instrument/dump_c$(OBJEXT) \
+  ../goto-instrument/goto_program2code$(OBJEXT) \
```
Comment (hannes-steffenhagen-diffblue) [inline src/goto-harness/Makefile:22]: these are all needed for `dump_c`

File: src/goto-harness/goto_harness_parse_options.cpp (hunk 12 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/goto_harness_parse_options.cpp]: @thk123 This snuck in from debugging earlier, it's not needed at all.

**Review by danpoe — APPROVED** [review]
LGTM

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]

File: src/goto-harness/recursive_initialization.h (hunk 16 above)
Comment (chrisr-diffblue) [reply src/goto-harness/recursive_initialization.h]: I would have thought that as this is an "implementation" prefix, it should be `__GOTO_HARNESS` ? (i.e. double underscore at the front) ?

File: src/util/type.cpp [hunk 19]
```diff
@@ -28,3 +28,9 @@ void type_with_subtypest::move_to_subtypes(typet &type)
   sub.push_back(static_cast<const typet &>(get_nil_irep()));
   sub.back().swap(type);
 }
+
+typet remove_const(typet type)
```
Comment (chrisr-diffblue) [inline src/util/type.cpp:32]: const argument? Or if this is modifying the argument, then why return ?

File: src/util/type.h (hunk 7 above)
Comment (chrisr-diffblue) [reply src/util/type.h:229]: Yeah, I'm not sure why this is a member function rather than a static?

**Review by chrisr-diffblue — COMMENTED** [review]
Mostly looks OK, but a couple of queries I'd like answered if possible?

File: src/util/type.cpp (hunk 19 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/type.cpp:32]: It's returning a copy with the modification. If the argument was const we'd have to make two copies, if it was const reference we'd still have to make a copy so we don't gain anything in either case.

File: src/util/type.h (hunk 7 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/type.h:229]: @chrisr-diffblue it is free standing, not a member. Thomas was suggesting moving it to goto harness where it was used at the time, but this is the sort of thing I could see being useful elsewhere as well and I figured we don't want to have too much proliferation of duplicated utility functions.

Comment (chrisr-diffblue) [reply src/util/type.h:229]: Ah, sorry - I failed to read the diff context properly :-)

File: src/util/type.cpp (hunk 19 above)
Comment (chrisr-diffblue) [reply src/util/type.cpp:32]: OK, I like the const just to make it explicit - and of course depending on how much this is called, an extra copy probably isn't the end of the world... But its a nitpick really on my part.

File: src/goto-harness/recursive_initialization.h (hunk 16 above)
Comment (hannes-steffenhagen-diffblue) [reply src/goto-harness/recursive_initialization.h]: We could certainly add a `__` in front.

**Review by chrisr-diffblue — APPROVED** [review]
LGTM

File: src/goto-harness/recursive_initialization.cpp [hunk 20]
```diff
@@ -774,7 +779,28 @@ code_blockt recursive_initializationt::build_struct_constructor(
...
+    // if the struct component is const we need to cast away the const
+    // for initialisation purposes.
+    // As far as I'm aware that's the closest thing to a 'correct' way
+    // to initialize dynamically allocated structs with const components
+    exprt component_initialisation_lhs = [&result, &component]() -> exprt {
+      auto member_expr = member_exprt{dereference_exprt{result}, component};
+      if(component.type().get_bool(ID_C_constant))
+      {
+        return dereference_exprt{
+          typecast_exprt{address_of_exprt{std::move(member_expr)},
```
Comment (smowton) [inline src/goto-harness/recursive_initialization.cpp:815]: Comment that this pile of stuff amounts to `*(X*)&field`, where `field` has type `const X*`

Comment (smowton) [reply src/goto-harness/recursive_initialization.cpp:815]: Also I take it you've tried using braced initialiser style for such things?

File: src/Makefile [hunk 21]
```diff
@@ -8,8 +8,8 @@ DIRS = analyses \
        goto-cc \
        goto-checker \
        goto-diff \
-       goto-harness \
        goto-instrument \
+       goto-harness \
```
Comment (smowton) [inline src/Makefile:12]: Sort order was correct before

**Review by smowton — APPROVED** [review]
Seems ok except the hundred or so tests that get marked KNOWNBUG here -- what was the memory-snapshot thingy that got trashed?

**Comment by NlightNFotis** [comment]
@smowton We are tracking the broken tests internally and through github issue https://github.com/diffblue/cbmc/issues/5351. The memory snapshot harness was one of the modes that the `goto-harness` tool supported.
